### PR TITLE
[BEAM-7305] Add first version of Hazelcast Jet based Java Runner

### PR DESCRIPTION
**Pull request by jbartok**
This is a first working version of a Java Runner based on Hazelcast Jet (https://github.com/hazelcast/hazelcast-jet). With it we have aimed for the batching (bounded data) scenario for now, but adding the missing pieces for unbounded data is something we fully intend to do. 

The functionality is relatively complete, see the **`validatesRunner`** task in the `:beam-runners-jet` module. I've excluded the Splittable DoFn tests completely, because we have not implemented that functionality yet and also the single test case for Impulse, because we have a problem with that related to parallelism. Beyond these though the results are: `235 tests completed, 10 failed, 2 skipped`. The 10 failed ones are also caused by missing features, like timer support, multi-graph support and such. Again, to be solved soon.

The **Nexmark testsuite** also runs, I'm using the following params: `-Pnexmark.runner=:beam-runners-jet -Pnexmark.args="--runner=JetRunner --suite=SMOKE --streaming=false --manageResources=false --monitorJobs=true --jetGroupName=jet"`

The last two Nexmark queries aren't being run successfully, but the cause of that seems to be external to our specific runner, a `NullPointerException` at `NexmarkQueryName.fromNumber(NexmarkQueryName.java:78)`. 

This is it. What we have done so far is mostly about pure functionality. We have not yet worked on performance (for example we don't do any pipeline surgery or fusing), which is far away from what standalone Jet is capable of, but I hope that this is still a worth-vile start. 

Lots and lots of things remain to be added and improved but I'm really curious what you think about it so far and would really appreciate your feedback.

R: @kennknowles 
R: @mxm 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kennknowles** [comment]
Nice!!

This is large enough I want to be clear: Have you filed an [ICLA](http://www.apache.org/licenses/#clas) with the ASF? Is there applicable CCLA? Anyone else contributing to this code?

**Comment by mxm** [comment]
Great to see this coming and already integrated with ValidatesRunner and Nexmark. I will definitely have a closer look the next days.

Usually it is enough for you to file the [ICLA](https://www.apache.org/licenses/icla.pdf), unless your company has to grant rights ([CCLA](https://www.apache.org/licenses/cla-corporate.txt)).

**Comment by jbartok** [comment]
FYI: we are working on the needed agreements, a CCLA will be required. I'll let you know once it's ready.

**Comment by jbartok** [comment]
I've submitted the CCLA.

**Comment by jbartok** [comment]
Hi guys. I've just added some more changes (a metrics related bug fix and generic handling for triggers). Wanted to ask if you had any time to look into this PR? I realise how much code this is, is there something I could do to facilitate your review? Walk you through the Jet part or anything else?

**Comment by mxm** [comment]
Apologies @jbartok, I'll take a look later today. Since you have integrated with ValidatesRunner and Nexmark, I don't think we need to review the Jet integration code in detail. Generally, we would expect members of the Jet community to maintain the Runner, or put differently, we want to welcome the Jet community inside the Beam community :) I just wanted to point that out, since it is also a possibility to maintain the Runner externally but still be listed as a Runner on the Beam website.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+/** Various common methods used by the Jet based runner. */
+public class Utils {
+
+  public static String getTupleTagId(PValue value) {
+    Map<TupleTag<?>, PValue> expansion = value.expand();
+    if (expansion.size() != 1) {
+      throw new RuntimeException(); // todo: Houston, we have a problem!
+    }
+    return expansion.keySet().iterator().next().getId();
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java]: Method body could be replaced by `Iterables.getOnlyElement(expansion.keySet())`;

File: runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      byte[] byteData = baos.toByteArray();
+      ByteArrayInputStream bais = new ByteArrayInputStream(byteData);
+      return (T) new ObjectInputStream(bais).readObject();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public static <T> byte[] encodeWindowedValue(
+      WindowedValue<T> windowedValue, Coder coder, ByteArrayOutputStream baos) {
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java]: Possibly you want to use `CoderUtils.encode(coder, value)`.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+      Coder<InputT> inputCoder,
+      Map<PCollectionView<?>, Coder<?>> sideInputCoders,
+      Map<TupleTag<?>, Coder<?>> outputCoders,
+      Coder<InputT> inputValueCoder,
+      Map<TupleTag<?>, Coder<?>> outputValueCoders,
+      Map<Integer, PCollectionView<?>> ordinalToSideInput,
+      String ownerId,
+      String stepId) {
+    this.pipelineOptions = pipelineOptions;
+    this.doFn = Utils.serde(doFn);
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java]: Why is this required? Maybe for local execution? Otherwise I'd assume that this would be cloned anyways when shipped to the cluster.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+/**
+ * Jet {@link com.hazelcast.jet.core.Processor} implementation for Beam's GroupByKeyOnly +
+ * GroupAlsoByWindow primitives.
+ *
+ * @param <K> key type of {@link KV} values from the output of this primitive
+ * @param <V> type of elements being windowed
+ */
+public class WindowGroupP<K, V> extends AbstractProcessor {
+
+  public static final byte[] EMPTY_BYTES = new byte[0];
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java]: Can be private?

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                    "Event type not supported in TestStream: " + e.getClass() + ", event: " + e);
+              }
+            })
+        .collect(toList());
+  }
+
+  @Override
+  public boolean complete() {
+    // todo: TestStream says it should cease emitting, but not stop after the items.
+    //   But I don't know how they end the job otherwise...
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java]: TestStream should ensure that each of the events are processed independently. That can mean waiting for an event to be completely processed before sending the next.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+                    new Throwable(
+                            WindowGroupP.class.getSimpleName()
+                                + " OUTPUT "
+                                + System.identityHashCode(WindowGroupP.this)
+                                + ", windowedValue = "
+                                + windowedValue)
+                        .printStackTrace();
+                  } // todo: remove
+                  // System.err.println(WindowGroupP.class.getSimpleName() + " OUTPUT ownerId = " +
+                  // ownerId + ", windowedValue = " + windowedValue); //useful for debugging
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java]: Remove?

File: runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  }
+
+  Vertex addVertex(String id, ProcessorMetaSupplier processorMetaSupplier) {
+    return dag.newVertex(id, processorMetaSupplier);
+  }
+
+  Vertex addVertex(String id, SupplierEx<Processor> processor) {
+    return dag.newVertex(id, processor)
+        .localParallelism(
+            localParallelism) // todo: quick and dirty hack for now, can't leave it like this
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java]: Why is this a hack?

File: runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        }
+
+        Coder edgeCoder = edgeCoders.get(edgeId);
+        if (edgeCoder == null) {
+          throw new RuntimeException("Oops!");
+        }
+
+        List<Vertex> destinationVertices =
+            edgeEndPoints.getOrDefault(edgeId, Collections.emptyList());
+        boolean sideInputEdge = edgeId.contains("PCollectionView"); // todo: this is a hack!
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java]: Perhaps clean this up?

File: runners/jet/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+
+  private static class ReadSourceTranslator<T>
+      implements JetTransformTranslator<PTransform<PBegin, PCollection<T>>> {
+
+    @Override
+    public Vertex translate(Pipeline pipeline, Node node, JetTranslationContext context) {
+      AppliedPTransform<PBegin, PCollection<T>, PTransform<PBegin, PCollection<T>>>
+          appliedTransform =
+              (AppliedPTransform<PBegin, PCollection<T>, PTransform<PBegin, PCollection<T>>>)
+                  node.toAppliedPTransform(pipeline);
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java]: You might want to change the caller to pass in the applied transform. Then you don't have to repeat this code for every transform translator.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  public static <T> ProcessorMetaSupplier supplier(
+      List<TestStream.Event<T>> events, Coder outputCoder) {
+    List<Object> serializableEvents = getSerializableEvents(events);
+    return ProcessorMetaSupplier.forceTotalParallelismOne(
+        ProcessorSupplier.of(() -> new TestStreamP(serializableEvents, outputCoder)));
+  }
+
+  private static <T> List<Object> getSerializableEvents(List<TestStream.Event<T>> events) {
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java]: You may want to encode the elements using the `valueCoder` in `TestStream`. Events do not have to be `Serializable`.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    public Object applyEx(byte[] b) throws Exception {
+      Object t = coder.decode(new ByteArrayInputStream(b)); // todo: decoding twice....
+      Object key = null;
+      if (t instanceof WindowedValue) {
+        t = ((WindowedValue) t).getValue();
+      }
+      if (t instanceof KV) {
+        key = ((KV) t).getKey();
+      }
+      return key == null ? "all" : key; // todo: why "all"?
```
Comment (mxm) [inline runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java]: I would also be curious :)

File: runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java (hunk 7 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java]: It's just a left-over TODO item, removing it in the next commit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java (hunk 1 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java]: True, thank you, will do so in the next commit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java (hunk 2 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java]: True, thank you, will do so in the next commit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java (hunk 4 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java]: Definitely, will be in the next commit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java (hunk 10 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java]: Ok, will be so in the next commit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java (hunk 9 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/JetTransformTranslators.java]: True, thank you, will do so in the next commit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java (hunk 3 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java]: Yes, for local execution. This is cloned when shipped to each member of a cluster, but then, inside one member there can be multiple processors using it. Leaving it as is.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java (hunk 10 above)
Comment (mxm) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java]: By "next commit", do you mean in this PR or in a follow-up?

Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java]: This PR, coming soon, just wanted to address all your concerns at once. At least the ones that I can. I'll push soon and also write a summary of future plans.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java (hunk 8 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java]: Ok, that's fair. I will improve on it, but atm I have no idea how. I've taken a note and will add it to the next PR we're planning to submit.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java (hunk 5 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java]: I understand what you are saying and indeed I've seen the need for such a behaviour myself, but I don't yet know how I could ensure such a behaviour when using Jet. Have taken a note and will look into before the next planned PR.

File: runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java (hunk 6 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java]: Ok, will do so in the next commit (this PR).

File: runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java (hunk 11 above)
Comment (jbartok) [reply runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java]: This "all" partitioning key is needed right now because we use partitioned edges all the time. But we shouldn't need to. I can't fix it immediately, but will look into it for the next PR.

**Comment by jbartok** [comment]
Hi Maximilian. First of all thank you very much for taking the time to review this. 

I've just pushed some changes, hopefully I've correctly addressed most of your concerns. 

Some of them I didn't, but I'm planning to do so in a follow-up Pull Request. We have been planning one anyways, because we want to make the Jet Runner more usable (and do some usability tests too, because it was only about having the necessary code until now) . For example add easy to use Gradle tasks for starting/stopping external Jet clusters, submitting Beam Pipelines to such clusters and so on. 

Btw. a question: once we have the usability part in place and we are confident that this runner is ok for Beam users to use, will it be ok to also submit an edit request for the Capability Matrix page, put ourselves "on the map", so to speak?

Best regards and thanks again for your help & time.

**Comment by mxm** [comment]
Thanks for the update. Of course, all Runners deserve an entrance in the capability matrix. Feel free to submit a follow-up PR for that.

I think this looks good to merge but I'd wait to see if @kennknowles wants to do a review as well.

**Comment by jbartok** [comment]
I've made the requested naming change.

**Comment by jbartok** [comment]
As far as I can tell the "Python_PVR_Flink" check failure is not caused by my changes...

**Comment by mxm** [comment]
Run Python_PVR_Flink PreCommit

**Comment by jbartok** [comment]
Hi Guys. Could you pls. merge this in, if everything is in order? I would like to open a new PR in which we would add a couple of missing pieces of code and start making the runner actually usable for running the Beam examples with. 

Sorry for pestering, I just want to not miss your next release. From what I understand from your build setup, once we are present in a released version it will be much simpler to use our runner for running the examples.

**Comment by jbartok** [comment]
I've looked at this conflict that popped up, but it's a bit tricky... The change on **master** that causes it is such that if I adapt to it, then exactly the part that I've used to name the runner "experimental" is what goes away... Should I instead rename "jet" into "experimental-jet"? 

And even if I ignore this aspect and try to adapt, there will still be work left on merge. Should I even attempt to change what's on the branch or will it be resolved during merge?

**Comment by mxm** [comment]
I agree, we should merge this. Everything is ok from my side and it looks like @kennknowles requested changes have been resolved. As for the release, we will have to cherry-pick this to the release branch because it has already been cut. Shouldn't be a problem though.

@jbartok Could you rebase the PR to master and get rid of any merge commits?

**Comment by mxm** [comment]
@jbartok I see that there are unrelated changes in some of the commits. Could you please remove those or simply squash all commits into a single commit? Will then merge.

**Comment by jbartok** [comment]
Done.

**Comment by mxm** [comment]
Thanks @jbartok! Could you assign this JIRA issue to yourself? https://jira.apache.org/jira/browse/BEAM-7305 If you do not have permissions, please ping me and provide your JIRA username.

You mentioned that you wanted to include the Runner in the release. Please open a PR against the `release-2.13.0` branch for that.

File: runners/jet-experimental/build.gradle [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+task validatesRunner {
+    group = "Verification"
+    description "Validates Jet runner"
+    dependsOn validatesRunnerBatch
+}
+
+spotless {
+    java {
+        paddedCell()
```
Comment (adude3141) [inline runners/jet-experimental/build.gradle]: Could you elaborate on why this is needed?

Currently, `spotlessJavaCheck`passes, but `spotlessApply`resolve some issues in codebase. Removing `paddedCell` will make `spotlessJavaCheck` fail.

So I am wondering, whether we could just remove this configuration?

Comment (kennknowles) [reply runners/jet-experimental/build.gradle]: Ah I just noticed this. Yes, we have hit bugs that slowed the whole community down because of paddedCell.

File: runners/jet-experimental/build.gradle [hunk 13]
```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
```
Comment (kennknowles) [inline runners/jet-experimental/build.gradle]: @adude3141 should there be some publishing config here to get the artifactId set up?

File: runners/jet-experimental/build.gradle (hunk 12 above)
Comment (jbartok) [reply runners/jet-experimental/build.gradle]: Hi Guys. I'm submitting a second pull request today with more changes related to jet runner and in that I'm removing this whole spotless block, it's not really needed.

Comment (jbartok) [reply runners/jet-experimental/build.gradle]: https://github.com/apache/beam/pull/8592

Comment (adude3141) [reply runners/jet-experimental/build.gradle]: @jbartok Nice, Thx!

File: runners/jet-experimental/build.gradle (hunk 13 above)
Comment (adude3141) [reply runners/jet-experimental/build.gradle]: Not quite sure, if I understand you correctly, @kennknowles 

I think, you are referring to archiveBaseName here. It is usually not necessary to setup on project level. Only if the default seems improper, we are ab le to overwrite.

E.g. in this case, the folder path is `runners:jet-experimental` which will result in artefact named something about `beam-runners-jet-experimemtal`. Now, we could decide that this is not wanted as after switching from experimental mode we would like to keep artifact named consistent. So we could add an
```
archivesBaseName = 'beam-runners-jet'
```

More likely in this though, would be thet we would like to drop that `experimental`on the folder name, i.e. move `runners/jet-experimental` to `runners/jet` and set

```
archivesBaseName = 'beam-runners-jet-experimental'
```
for the time being.

(Hope I got it right)

Comment (kennknowles) [reply runners/jet-experimental/build.gradle]: OK, I see. I did not quite understand that the archive base name was so well automated. That is great.
